internal/sets/mapset.go

```diff
@@ -0,0 +1,153 @@
+package sets
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"golang.org/x/exp/maps"
+)
+
+// Similar to the concept of a [legal peppercorn](https://en.wikipedia.org/wiki/Peppercorn_(law)), this instance of
+// nothingness is required in order to transact with Go's map[T]struct{} idiom.
+var peppercorn = struct{}{}
+
+// MapSet is a struct that adds some convenience to the otherwise cumbersome map[T]struct{} idiom used in Go to
+// implement sets of comparable types.
+type MapSet[T comparable] struct {
+	m map[T]struct{}
+}
+
+// NewMapSet returns a MapSet ready for use. Optionally, a desired size for the MapSet can be passed as an argument,
+// as in the argument to make() for a map type.
+func NewMapSet[T comparable](args ...int) MapSet[T] {
+	if len(args) > 1 {
+		panic(fmt.Sprintf("too many arguments passed to NewMapSet(). got: %v, expected 0 or 1", len(args)))
+	}
+
+	var size int
+	if len(args) == 1 {
+		size = args[0]
+	}
+
+	var m map[T]struct{}
+	if size > 0 {
+		m = make(map[T]struct{}, size)
+	}
+
+	return MapSet[T]{m: m}
+}
+
+// NewMapSetFromSlice creates a MapSet of size len(items) and calls AddSlice(items) on it.
+func NewMapSetFromSlice[T comparable](items []T) MapSet[T] {
+	h := NewMapSet[T](len(items))
+	h.AddSlice(items)
+	return h
+}
+
+// Add an item to the set. Returns true if the item did not exist in the set.
+func (h *MapSet[T]) Add(item T) bool {
+	if h.m == nil {
+		h.m = map[T]struct{}{}
+	}
+
+	_, exists := h.m[item]
+	h.m[item] = peppercorn
```

Comment:
Maybe do an `if` around the exists and just return true.  Then return false in the other case.